### PR TITLE
add support for loading data from yaml files [#444]

### PR DESCRIPTION
**Pull request by P-Andersson**
This change adds support for loading YAML data files, alongside json, csv and XML. Mappings whose keys cannot be easily converted to strings are filtered out entirely, in order to preserve the interface of the JSON loader. This should still be fine for most inputs one can expect I think. But feel free to comment!

### CODE REVIEW

File: library/src/compute/data.rs [hunk 1]
```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
...
+/// converted into Typst dictionaries, and YAML sequences will be converted into
+/// Typst arrays. Strings and booleans will be converted into the Typst
+/// equivalents, null-values (`null`, `~` or empty ``) will be converted into 
+/// `{none}`, and numbers will be converted to floats or integers depending on 
+/// whether they are whole numbers.
+///
+/// Note that mapping keys that are not a string, a number, a boolean or a 
+/// null-value will cause the entry to be discarded. Otherwise it will be 
+/// converted to a string.
+/// Custom YAML tags are ignored, though the loaderd value will still be
```
Comment (laurmaedje) [inline library/src/compute/data.rs:225]:
```suggestion
/// Custom YAML tags are ignored, though the loaded value will still be 
```

File: library/src/compute/data.rs [hunk 2]
```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
...
+pub fn yaml(
+    /// Path to a JSON file.
+    path: Spanned<EcoString>,
+) -> Value {
+    let Spanned { v: path, span } = path;
+    let path = vm.locate(&path).at(span)?;
+    let data = vm.world().file(&path).at(span)?;
+    let value: serde_yaml::Value =
+        serde_yaml::from_slice(&data).map_err(format_yaml_error).at(span)?;
+        convert_yaml(value)
```
Comment (laurmaedje) [inline library/src/compute/data.rs:259]:
```suggestion
    convert_yaml(value)
```

File: library/src/compute/data.rs [hunk 3]
```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
...
+            None => Value::Float(v.as_f64().unwrap_or(f64::NAN)),
+        },
+        serde_yaml::Value::String(v) => Value::Str(v.into()),
+        serde_yaml::Value::Sequence(v) => {
+            Value::Array(v.into_iter().map(convert_yaml).collect())
+        }
+        serde_yaml::Value::Mapping(v) => Value::Dict(
+            v.into_iter()
+                .map(|(key, value)| (convert_yaml_key(key), convert_yaml(value)))
+                .filter(|(key, _)| key.is_some())
```
Comment (laurmaedje) [inline library/src/compute/data.rs:278]: could probably use filter_map with an inner map over the key

File: library/src/compute/data.rs [hunk 4]
```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
...
+                .collect(),
+        )
+    }
+}
+
+/// Converts an arbitary YAML mapping key, value into a String, 
+/// or None if not possible
+fn convert_yaml_key(key: serde_yaml::Value) -> Option<Str> {
+    match key {
+        serde_yaml::Value::Null => Some(Str::from("none")),
```
Comment (laurmaedje) [inline library/src/compute/data.rs:289]: hmmm, should this be `"none"` or `"null"`? should we just disallow all non-string keys or are they common?

**Review by laurmaedje — CHANGES_REQUESTED** [review]
Seems like a good addition!

File: library/src/compute/data.rs (hunk 4 above)
Comment (P-Andersson) [reply library/src/compute/data.rs:289]: Value::None gets converted to "none" by default in Typst so it should be that for the sake of concistency. But you are right that it's likely not very common and will conflict with the "none" string regardless, so we can probable sort it out.

Comment (laurmaedje) [reply library/src/compute/data.rs]: I think I'd prefer to just filter out all non-string keys. Typst dictionaries might gain support for non-string keys in the future and then we could add this.

File: library/src/compute/data.rs [hunk 5]
```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
...
+///        #author
+///        #list(..works.map((work => [#work.title (#work.published)])))
+///    ]
+///]
+///#bookshelf(yaml("scifi-authors.yaml"))
+/// ```
+///
+/// Display: YAML
+/// Category: data-loading
+/// Returns: array, value or dictionary
```
[the remark this answers is not in the document]
Comment (laurmaedje) [reply library/src/compute/data.rs:251]: This is still unfixed.

File: library/src/compute/data.rs (hunk 4 above)
Comment (P-Andersson) [reply library/src/compute/data.rs]: Fair enough; I'll change and document that behavior

**Pull request merged**